### ISSUE TITLE
Issues with National Library of Israel site

### ISSUE THREAD

**Issue by Shanzy** [lovasoa/dezoomify-rs#107]
I was recently looking at this book, on the National Library of Israel's site: https://web.nli.org.il/sites/NLI/English/digitallibrary/pages/viewer.aspx?presentorid=MANUSCRIPTS&docid=PNX_MANUSCRIPTS990001748120205171-1#|FL81030630

The site utilizes IIIF to zoom, thus the corresponding Json file for the first page is: https://iiif.nli.org.il/IIIFv21/FL81030630/info.json

This works fine on the dezoomify website, but for some reason when I run dezoomify-rs from the command line, the images aren't stitched together properly - there are black boxes in between each photo like a chessboard:
https://user-images.githubusercontent.com/78405487/106564534-56289080-64fb-11eb-8476-27d5f7f892e0.jpg

Something seems to be different on dezoomify-rs from the website that is causing it to break.

**Comment by lovasoa** [comment]
The website itself doesn't seem to work:
![image](https://user-images.githubusercontent.com/552629/106650553-54aaa700-6593-11eb-8176-06cc848f9dff.png)

The website's implementation of IIIF seems to be buggy. For instance, the following URL: 

https://iiif.nli.org.il/IIIFv21/FL81030630/0,1024,1024,1024/1024,1024/0/default.jpg

should return a 1024x1024 image according to the [IIIF specification](https://iiif.io/api/image/3.0/), but it returns:
 
![](https://iiif.nli.org.il/IIIFv21/FL81030630/0,1024,1024,1024/1024,1024/0/default.jpg)

which is a 512x512 image

**Comment by Shanzy** [comment]
> The website itself doesn't seem to work:
> ![image](https://user-images.githubusercontent.com/552629/106650553-54aaa700-6593-11eb-8176-06cc848f9dff.png)
> 
> And the website's implementation of IIIF seems to be buggy. For instance, the following URL:
> 
> https://iiif.nli.org.il/IIIFv21/FL81030630/0,1024,1024,1024/1024,1024/0/default.jpg
> 
> should return a 1024x1024 image according to the [IIIF specification](https://iiif.io/api/image/3.0/), but it returns:
> 
> ![](https://camo.githubusercontent.com/a6faf95284a51783fd923e8ba2018aec6c27715f0b4fd48f22a13b8e9eb589c3/68747470733a2f2f696969662e6e6c692e6f72672e696c2f494949467632312f464c38313033303633302f302c313032342c313032342c313032342f313032342c313032342f302f64656661756c742e6a7067)
> 
> which is a 512x512 image

Thanks for the help! The link can be wonky - if you want to get it, you can go to https://web.nli.org.il/sites/NLIS/en/ManuScript/Pages/Item.aspx?ItemID=PNX_MANUSCRIPTS990001748120205171 and click the picture on the right.

Are there any workarounds I could try for downloading these photos using dezoomify-rs despite the site-side bug? I'm also wondering why IIIF seems to work fine on dezoomify itself - it's only when using the command line that I get fractured images.

**Comment by lovasoa** [comment]
You can save the info.json file locally, fix it, and then point dezoomify-rs to it.
The fix is : 

```json
{
  "@context" : "http://iiif.io/api/image/2/context.json",
  "@id" : "https://iiif.nli.org.il/IIIFv21/FL81030630",
  "protocol" : "http://iiif.io/api/image",
  "width" : 9616,
  "height" : 5760,
  "sizes" : [ {
    "width" : 75,
    "height" : 45
  }, {
    "width" : 150,
    "height" : 90
  }, {
    "width" : 300,
    "height" : 180
  } ],
  "tiles" : [ {
    "width" : 576,
    "height" : 576,
    "scaleFactors" : [ 1, 2, 4, 8, 16, 32, 64, 128 ]
  } ],
  "profile" : [ "http://iiif.io/api/image/2/level1.json", {
    "formats" : [ "jpg" ],
    "qualities" : [ "native", "color", "gray" ],
    "supports" : [ "regionByPct", "regionSquare", "sizeByForcedWh", "sizeByWh", "sizeAboveFull", "rotationBy90s", "mirroring" ],
    "maxWidth" : 526,
    "maxHeight" : 526,
    "maxArea" : 111111
  } ],
  "license" : [ "http://web.nli.org.il/sites/NLI/English/library/Pages/terms.aspx" ],
  "logo" : [ "http://web.nli.org.il/sites/NLI/Hebrew/library/aboutus/now/projects/IAN/PublishingImages/NLI_logo_transparent_small.png" ],
  "attribution" : [ "[{\"@value\":\"The National Library of Israel\",\"@language\":\"en\"},{\"@value\":\"הספרייה הלאומית\",\"@language\":\"he\"}]" ]
}
```


![dz](https://user-images.githubusercontent.com/552629/106659051-81b08700-659e-11eb-8ae3-82d0c2897876.jpg)

**Comment by lovasoa** [comment]
Hmm, actually, looking at the info file more carefully, it looks like it specifies a maxWidth which is inferior to the specified tile width. We could add support for that.

**Comment by lovasoa** [comment]
dezoomify has a special logic encoded, where it first loads the first tile to detect the image size: https://github.com/lovasoa/dezoomify/blob/master/dezoomers/iiif.js#L87-L96 . That's why it works on dezoomify and not dezoomify-rs

**Comment by Shanzy** [comment]
All the properties of the Json file and the URL are discussed [here](https://iiif.io/api/image/3.0/)
I think the issue is that the maxheight and maxwidth are both 526, but the tiles are 1024 x 1024. Normally this isn't allowed, but they also have an extra functionality feature called sizeabovefull, which allows the client to request images larger than the max values. If you notice that sizeabovefull is not one of the extra functionality features listed in the above link, it's because it was phased out years ago due to security issues, but it's in an earlier version [here](https://iiif.io/api/image/2.0/). Maybe all that crap is getting in the way, since NLI never updated.

**Comment by lovasoa** [comment]
Yes, their file doesn't seem to be compliant with either IIIFv2 or IIIFv3. None seems to allow a `maxWidth` property in a `profile` object. v3 has `maxWidth`, but at the top level. Do you speak hebrew ? Could you report the bug to them ?

**Comment by Shanzy** [comment]
> Yes, their file doesn't seem to be compliant with either IIIFv2 or IIIFv3. None seems to allow a `maxWidth` property in a `profile` object. v3 has `maxWidth`, but at the top level. Do you speak hebrew ? Could you report the bug to them ?

I do, so I'll definitely give them a shout. 

In the meantime adding support for the maxWidth issue would be amazing if you can ever find the time - I assume NLI isn't the only buggy site, so it would prevent a lot of similar issues. Thanks again for all the help, I really appreciate it!

**Comment by dovbear75** [comment]
Hello,
in last weeks I cann’t download the images from the National Library of Israel. Originally I copied a link from Network Cache Storage, e. g. https://iiif.nli.org.il/IIIFv21/FL53708510/1024,0,1024,1024/1024,/0/default.jpg. 
Now it doesn’t work, the second possibility via dezoomify plugin (in Firefox) offers me the following link for Dezoomify: https://web.nli.org.il/sites/NLI/Hebrew/digitallibrary/pages/viewer.aspx?&presentorid=MANUSCRIPTS&docid=PNX_MANUSCRIPTS990000637470205171-1#|FL53708510
Again without any success, where is a problem, please? 
Thank you for help and best wishes,
dovbear75

**Comment by lovasoa** [comment]
None of your links seem to contain an image.

<img width="547" alt="image" src="https://user-images.githubusercontent.com/552629/119956452-692b0900-bfa1-11eb-9def-36d28add5518.png">

<img width="1680" alt="image" src="https://user-images.githubusercontent.com/552629/119956491-6f20ea00-bfa1-11eb-84f4-2d508f9580c0.png">

**Comment by lovasoa** [comment]
It doesn't seem to work for me either

<img width="1158" alt="image" src="https://user-images.githubusercontent.com/552629/119989274-47447d00-bfc7-11eb-8825-305f45dcf3f0.png">

**Comment by dovbear75** [comment]
It’s strange, because all the year I worked with your perfect dezoomify
application without any problem and now I can't find any way to solve it.
I hope you’ll find some solution in close future. ;-)
Best wishes and good luck,
Daniel (dovbear)

pá 28. 5. 2021 v 15:14 odesílatel Ophir LOJKINE ***@***.***>
napsal:

**Comment by lovasoa** [comment]
I can't find a solution if I can't see the problem. There is no zoomable image that appears for me on the link you gave. I tried both in chrome and firefox.

**Comment by dovbear75** [comment]
Dear Ophir,

It’s weird. I used your application many times and no problem. Just it
begined the problems with downloading the images a few weeks ago.
E. g.: a regular link for studying the zoomable pictures:
https://web.nli.org.il/sites/NLI/English/digitallibrary/pages/viewer.aspx?&presentorid=MANUSCRIPTS&docid=PNX_MANUSCRIPTS990000637470205171-1#|FL53708507
[image: screenshot_NLI.png]
or clicking on the right picture at this website:
https://web.nli.org.il/sites/NLIS/en/ManuScript/Pages/Item.aspx?ItemID=PNX_MANUSCRIPTS990000637470205171

Do you can see it now, please?
Thanks for understanding and best wishes,
Daniel (dovbear)

pá 28. 5. 2021 v 20:01 odesílatel Ophir LOJKINE ***@***.***>
napsal:

**Comment by Shanzy** [comment]
Update - the links to the scans at the NLI are kind of wonky, so here’s a better example:
Go to https://web.nli.org.il/sites/NLIS/en/ManuScript/Pages/Item.aspx?ItemID=PNX_MANUSCRIPTS990001516890205171 and hit the picture captioned “link to additional scans.” That should bring you to the scan.
The corresponding Json file for the first image of this scan is https://iiif.nli.org.il/IIIFv21/FL38636645/info.json (you can see “FL38636645” at the end of the URL for the scan when the first image is loaded). This link loads fine in the browser. What’s going wrong is that when Dezomify is run, it’s messing up and not using the right numbers for the actual components of the picture. For example, the first url it tried to download is:
https://iiif.nli.org.il/IIIFv21/FL38636645/3072,0,1024,1024/1024,1024/0/default.jpg 
When in fact this URL does not exist (resulting in a 500 error). The real link is: 
https://iiif.nli.org.il/IIIFv21/FL38636645/3072,0,1024,1024/1024,/0/default.jpg 
Which you can see in the source files of the image itself. 
The second link it tries to download is:
https://iiif.nli.org.il/IIIFv21/FL38636645/4096,1024,640,224/640,224/0/default.jpg 
But there’s nothing like this in the image files themselves.
Basically, something is causing dezoomify to use the wrong numbers for the URL it wants to download (is this specific to the NLI?). I know the Json files themselves for the images on the NLI site are messed up a bit, so maybe this is related. It results in a blank image every time.

I’m especially noticing how it adds an extra number after the comma in between the second and third slashes. So while the source image is .../1024,/0... Dezoomify will search for .../1024,224/0... 
I wonder if there is a way to download and edit the Json file to fix this manually. If there is I could run a script to do it.

**Comment by lovasoa** [comment]
The problem is not that dezoomify requests the wrong URL, it is that the server refuses to serve requests if the `Referer` http header is not `https://web.nli.org.il`.

![image](https://user-images.githubusercontent.com/552629/122199388-4900b280-ce9a-11eb-880c-50f6eae710f0.png)

You can work around that easily with [dezoomify-rs](https://dezoomify-rs.ophir.dev/) : 

``` 
❯  dezoomify-rs "https://iiif.nli.org.il/IIIFv21/FL38636658/info.json" -H "Referer: https://web.nli.org.il"

Found the following zoom levels:
 0. FL38636658 ( 5888 x  4240 pixels,    30 tiles)
 1. FL38636658 ( 2944 x  2120 pixels,     9 tiles)
 2. FL38636658 ( 1472 x  1060 pixels,     4 tiles)
 3. FL38636658 (  736 x   530 pixels,     1 tiles)
 4. FL38636658 (  368 x   265 pixels,     1 tiles)
 5. FL38636658 (  184 x   132 pixels,     1 tiles)
 6. FL38636658 (   92 x    66 pixels,     1 tiles)
 7. FL38636658 (   46 x    33 pixels,     1 tiles)
```

**Comment by lovasoa** [comment]
And the latest development builds support the maxWidth feature used in the NLI.

https://github.com/lovasoa/dezoomify-rs/releases/tag/build-efa91756b53c40dcd13ef6eb2ad0dc37f9e79c9c


@Shanzy Could you test it and give feedback ?

**Comment by Shanzy** [comment]
> And the latest development builds support the maxWidth feature used in the NLI.
> 
> https://github.com/lovasoa/dezoomify-rs/releases/tag/build-efa91756b53c40dcd13ef6eb2ad0dc37f9e79c9c
> 
> @Shanzy Could you test it and give feedback ?

I downloaded the new build and tried it on several manuscripts from the NLI site. Works like a charm - there are no longer any black tiles in middle of the image like before, which means I no longer have to modify the Json files with a script. As I’ve only tried it on the zoomer offered by the NLI, I don’t know whether the new code will interfere with any of dezoomify-rs’ regular abilities, but I doubt it. The update seems to accomplish exactly what it was created for. Making this build (it’s a pre-release at the moment, right?) the regular, downloadable version would be great, If I notice something wrong later, I’ll be sure to let you know. As always, thanks so much!
